### PR TITLE
Add support for protobuf maps

### PR DESCRIPTION
**Pull request by andremarianiello**
First part of https://github.com/centrifuge/precise-proofs/issues/31

### CODE REVIEW

File: proofs/salts.go [hunk 1]
```diff
@@ -53,6 +55,40 @@ func FillSalts(dataMessage, saltsMessage proto.Message) (err error) {
 	return
 }
 
+func handleFillSaltsMap(saltsType reflect.Type, saltsField reflect.Value, valueField reflect.Value) (err error) {
+	if saltsType.Kind() != reflect.Map || reflect.TypeOf(valueField.Interface()).Kind() != reflect.Map {
```
[the remark this answers is not in the document]
Comment (andremarianiello) [reply proofs/salts.go]: Added tests for this. It is covered as well as the slice case now.

File: proofs/property.go [hunk 2]
```diff
@@ -165,3 +186,54 @@ func AsBytes(propName proofspb.PropertyName) []byte {
...
+			case `\`:
+				return `\\`
+			case `.`:
+				return `\.`
+			case `[`:
+				return `\[`
+			case `]`:
+				return `\]`
+			}
+			return "???"
```
[the remark this answers is not in the document]
Comment (andremarianiello) [reply proofs/property.go]: Added the panic.

File: proofs/property.go [hunk 3]
```diff
@@ -84,16 +87,34 @@ func (n Property) FieldPropFromTag(protobufTag string) (Property, error) {
...
-// ElemProp takes a repeated field index and returns a child Property representing that element of the repeated field
-func (n Property) ElemProp(i FieldNum) Property {
+// SliceElemProp takes a repeated field index and returns a child Property representing that element of the repeated field
+func (n Property) SliceElemProp(i FieldNum) Property {
 	return Property{
 		Parent:     &n,
 		Text:       fmt.Sprintf("%d", i),
 		Nums:       []FieldNum{i},
-		NameFormat: SliceElemFormat,
+		NameFormat: ElemFormat,
```
[the remark this answers is not in the document]
Comment (andremarianiello) [reply proofs/property.go]: Not sure what you mean here. Can you explain?

[the remark this answers is not in the document]
Comment (lucasvo) [reply proofs/property.go]: I think the "NameFormat" name is perhaps not the best, it wasn't entirely clear what it is used for but looking at the code again, it makes sense. It is used when creating new properties from a parent one.

File: proofs/flatten.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		f.appendLeaf(*prop, valueString, saltValue.Interface().([]byte), []byte{}, false)
+		return nil
+	}
+
+	// handle generic recursive cases
+	switch value.Kind() {
+	case reflect.Ptr:
+		return f.handleValue(prop, value.Elem(), saltValue.Elem(), reflect.Value{}, fieldDescriptor)
+	case reflect.Struct:
+
```
Comment (lucasvo) [inline proofs/flatten.go]: remove space

**Pull request merged**